---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE}
library(PCAtools)
library(fclust)
library(readxl)
require(dplyr)
require(readr)
require(TxDb.Mmusculus.UCSC.mm10.knownGene)
require(Organism.dplyr)
```

```{r}
metaData <- read_xlsx("7342_MM_Samples_IDs.xlsx")
colnames(metaData) <- c("sampleid", "Description")
metaData$timepoint <- sapply(metaData$Description, function(x){strsplit(x, split = "\\s")[[1]][1]})
metaData$treatment <- sapply(metaData$Description, function(x){strsplit(x, split = "\\s")[[1]][2]})
metaData$treatment[c(1:8, 21:24, 37:40)] <- "Vehicle"
metaData$treatment[c(17:20,  33:36, 49:52)] <- "No_Q"

metaData$group <- paste0(metaData$treatment, ".", metaData$timepoint)
#metaData$group <- c(1,1,1,1,rep(2,16), rep(3,16),rep(4,16))

bam_files <- c("7342-MM-1.filtered.bam","7342-MM-2.filtered.bam","7342-MM-3.filtered.bam",
               "7342-MM-4.filtered.bam","7342-MM-5.filtered.bam","7342-MM-6.filtered.bam",
               "7342-MM-7.filtered.bam","7342-MM-8.filtered.bam","7342-MM-9.filtered.bam",
               "7342-MM-10.filtered.bam","7342-MM-11.filtered.bam","7342-MM-12.filtered.bam",
               "7342-MM-13.filtered.bam","7342-MM-14.filtered.bam","7342-MM-15.filtered.bam",
               "7342-MM-16.filtered.bam","7342-MM-17.filtered.bam","7342-MM-18.filtered.bam",
               "7342-MM-19.filtered.bam","7342-MM-20.filtered.bam","7342-MM-21.filtered.bam",
               "7342-MM-22.filtered.bam","7342-MM-23.filtered.bam","7342-MM-24.filtered.bam",
               "7342-MM-25.filtered.bam","7342-MM-26.filtered.bam","7342-MM-27.filtered.bam",
               "7342-MM-28.filtered.bam","7342-MM-29.filtered.bam","7342-MM-30.filtered.bam",
               "7342-MM-31.filtered.bam","7342-MM-32.filtered.bam","7342-MM-33.filtered.bam",
               "7342-MM-34.filtered.bam","7342-MM-35.filtered.bam","7342-MM-36.filtered.bam",
               "7342-MM-37.filtered.bam","7342-MM-38.filtered.bam","7342-MM-39.filtered.bam",
               "7342-MM-40.filtered.bam","7342-MM-41.filtered.bam","7342-MM-42.filtered.bam",
               "7342-MM-43.filtered.bam","7342-MM-44.filtered.bam","7342-MM-45.filtered.bam",
               "7342-MM-46.filtered.bam","7342-MM-47.filtered.bam","7342-MM-48.filtered.bam",
               "7342-MM-49.filtered.bam","7342-MM-50.filtered.bam",
               "7342-MM-51.filtered.bam","7342-MM-52.filtered.bam")

metaData$bamfile <- bam_files


#CD8_RNAseq_GLUTAMINE_METABOLISM_SAMPLES <- metaData[,c("sampleid", "timepoint", "group", "bamfile")]
#CD8_RNAseq_GLUTAMINE_METABOLISM_SAMPLES <- metaData
#use_data(CD8_RNAseq_GLUTAMINE_METABOLISM_SAMPLES, overwrite = TRUE)

counts <- read.table(file = "CD8_RNAseq_counts.txt", sep = "\t", header = TRUE, skip = 1)
colnames(counts) <- c("Geneid", "Chr", "Start", "End", "Strand", "Length", metaData$sampleid)
Geneid <- counts$Geneid

CD8_RNAseq_GLUTAMINE_METABOLISM_FEATURE_COUNT <- counts

## raw counts
#use_data(CD8_RNAseq_GLUTAMINE_METABOLISM_FEATURE_COUNT)

counts <- read.table(file = "CD8_RNAseq_counts.txt", sep = "\t", header = TRUE, skip = 1)
colnames(counts) <- c("Geneid", "Chr", "Start", "End", "Strand", "Length", metaData$sampleid)

src <- src_ucsc(organism = "mouse", genome = "mm10")
IdList <- sapply(counts$Geneid, function(x){strsplit(x, "\\.")}[[1]][[1]])
counts$Geneid <- IdList
sym <- select(src, keys = IdList, columns = c("entrez", "symbol", "ensembl", "gene_chrom", "gene_start", "gene_end"),
              keytype = "ensembl")
colnames(sym) <- c("Geneid", "entrez", "symbol", "chr", "start", "end")

counts <- left_join(counts, sym, by = "Geneid")
counts <- counts[!is.na(counts$entrez), ]
counts <- counts[!duplicated(counts$entrez), ]
rownames(counts) <- counts$entrez
CD8_RNAseq_GLUTAMINE_METABOLISM_COUNTS <- counts


ctx <- as.matrix(CD8_RNAseq_GLUTAMINE_METABOLISM_COUNTS[, 7:58])
rownames(ctx) <- CD8_RNAseq_GLUTAMINE_METABOLISM_COUNTS$entrez
metaData <- data.frame(metaData)
rownames(metaData) <- colnames(ctx)
ctx <- log2(ctx+0.1)
p <- pca(t(ctx), metadata = NULL, scale = TRUE, rank = 20, removeVar = 0.1)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
biplot(p, x = "PC2", y = "PC3", showLoadings = TRUE, ntopLoadings = 52, maxoverlapsConnectors = 52,
       labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
```


```{r}
ctx_t <- t(ctx)
fc <- Fclust(ctx_t, k = 6, type="standard",ent=FALSE,noise=FALSE,stand=1,distance=FALSE)
```

